plugin/sampling/strategystore/static/strategy_store.go

```diff
@@ -29,31 +32,102 @@ import (
 )
 
 type strategyStore struct {
+	// to allow concurrent update of sampling strategies
+	lock sync.RWMutex
+
 	logger *zap.Logger
 
 	defaultStrategy   *sampling.SamplingStrategyResponse
 	serviceStrategies map[string]*sampling.SamplingStrategyResponse
 }
 
+func (h *strategyStore) loadAndParseStrategies(strategiesFile string) error {
+	s, err := loadStrategies(strategiesFile)
+	if err != nil {
+		h.logger.Warn("using the last saved configuration for sampling strategies.", zap.Error(err))
+		return err
+	}
+
+	h.parseStrategies(s)
+	return nil
+}
+
 // NewStrategyStore creates a strategy store that holds static sampling strategies.
 func NewStrategyStore(options Options, logger *zap.Logger) (ss.StrategyStore, error) {
 	h := &strategyStore{
 		logger:            logger,
 		serviceStrategies: make(map[string]*sampling.SamplingStrategyResponse),
 	}
-	strategies, err := loadStrategies(options.StrategiesFile)
-	if err != nil {
+
+	// Read strategies
+	if err := h.loadAndParseStrategies(options.StrategiesFile); err != nil {
 		return nil, err
 	}
-	h.parseStrategies(strategies)
+
+	// Watch strategies file for changes.
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		logger.Error("failed to create a new watcher for the sampling strategies file", zap.Error(err))
+	}
+
+	go h.runWatcherLoop(watcher, options.StrategiesFile)
+
+	if err := watcher.Add(options.StrategiesFile); err != nil {
+		logger.Error("error adding watcher to file", zap.String("file", options.StrategiesFile), zap.Error(err))
+	} else {
+		logger.Info("watching", zap.String("file", options.StrategiesFile))
+	}
+
+	dir := filepath.Dir(options.StrategiesFile)
+	err = watcher.Add(dir)
+	if err := watcher.Add(dir); err != nil {
+		h.logger.Error("error adding watcher to dir", zap.String("dir", dir), zap.Error(err))
+	} else {
+		h.logger.Info("watching", zap.String("dir", dir))
+	}
+
 	return h, nil
 }
 
+func (h *strategyStore) runWatcherLoop(watcher *fsnotify.Watcher, strategiesFile string) {
+	for {
+		select {
+		case event := <-watcher.Events:
```

Comment:
Could we not simplify this by handling all event? E.g.
```go
case event := <-watcher.Events:
    if event.Op == fsnotify.Remove {
        // resubscribe to handle k8s use case
    }
    h.loadAndParseStrategies(strategiesFile)
    continue
```

Reply:
The current logic is actually very close to this 

1. Remove events could occur in a non k8s environment. In which case we need to handle only write events, hence the following block-
```			
if event.Op&fsnotify.Write == fsnotify.Write {
	h.loadAndParseStrategies(strategiesFile)
}
```

2. If they occur in a k8s environment, we need to reload the file strategies file hence the first part

Reply:
Just a note: the second case isn't specific to Kubernetes, but it's how Kubernetes mounts those files in the local file system (with two symlinks at different levels). The same would happen on non-Kubernetes environments as well, if people use symlinks on directories pointing to the "current" version of a config map.

Reply:
my point is that you want special handing for Remove by re-subscribing, and all events should result in parsing the file. Right now the code ignores several events altogether.

Reply:
> Right now the code ignores several events altogether.

?